src/main/python/org/broadinstitute/hellbender/tools/gcnv/gcnvkernel/tasks/inference_task_base.py

```diff
@@ -0,0 +1,757 @@
+
+        if 'elbo_normalization_factor' in kwargs.keys():
+            self.elbo_normalization_factor = kwargs['elbo_normalization_factor']
+        else:
+            self.elbo_normalization_factor = 1.0
+
+        if 'advi_task_name' in kwargs.keys():
+            self.advi_task_name = kwargs['advi_task_name']
+        else:
+            self.advi_task_name = "ADVI"
+
+        if 'sampling_task_name' in kwargs.keys():
+            self.sampling_task_name = kwargs['sampling_task_name']
+        else:
+            self.sampling_task_name = "sampling"
+
+        if 'calling_task_name' in kwargs.keys():
+            self.calling_task_name = kwargs['calling_task_name']
+        else:
+            self.calling_task_name = "calling_task_name"
+
+        self._t0 = None
+        self._t1 = None
+        self.elbo_hist: List[float] = []
+        self.rls_elbo_hist: List[float] = []
+        self.snr_hist: List[float] = []
+        self.i_epoch = 1
+        self.i_advi = 1
+        self.calling_hist: List[Tuple[int, bool, bool]] = []
+        self.previous_sampling_rounds = 0
+        self.latest_caller_update_summary: Optional[CallerUpdateSummary] = None
+        self.tqdm_out = LoggerTQDMAdapter(_logger)
+
+    @abstractmethod
+    def disengage(self):
+        raise NotImplementedError
+
+    def engage(self):
+        try:
+            all_converged = False
+            while self.i_epoch <= self.hybrid_inference_params.max_training_epochs:
+                _logger.debug("Starting epoch {0}...".format(self.i_epoch))
+                converged_continuous = self._update_continuous_posteriors()
+                all_converged = converged_continuous
+                if self.sampler is not None:
+                    converged_sampling = self._update_log_emission_posterior_expectation()
+                    all_converged = all_converged and converged_sampling
+                if self.caller is not None:
+                    converged_discrete = self._update_discrete_posteriors()
+                    all_converged = all_converged and converged_discrete
+                self.i_epoch += 1
+                if all_converged and not self._premature_convergence():
+                    break
+            if all_converged:
+                _logger.info("Inference task completed successfully and convergence achieved.")
+            else:
+                _logger.warning("Inference task completed successfully but convergence not achieved.")
+        except KeyboardInterrupt:
+            pass
+
+    def _log_start(self, task_name: str, i_epoch: int):
+        self._t0 = time.time()
+        _logger.debug("Starting {0} for epoch {1}...".format(task_name, i_epoch))
+
+    def _log_stop(self, task_name: str, i_epoch: int):
+        self._t1 = time.time()
+        _logger.debug('The {0} for epoch {1} successfully finished in {2:.2f}s'.format(
+            task_name, i_epoch, self._t1 - self._t0))
+
+    def _log_interrupt(self, task_name: str, i_epoch: int):
```

Comment:
Just curious, why do you take special precautions for `KeyboardInterrupt`?  Is this a TQDM thing?

Reply:
It is very useful for interactive mode. I don't think it would have any effect if the scripts are run via `PythonScriptExecutor` since `stdin` is not piped (is it @cmnbroad?)

Reply:
Right. Technically I think you could write to stdin and the underlying process controller would honor it, but the PythonScriptExecutor doesn't rely on it (since you always run a script, module, or command line command). The StreamingPythonScriptExecutor does so though.

Reply:
@cmnbroad Perfect, thank you!